src/prefect/_internal/concurrency/workers.py

```diff
@@ -0,0 +1,172 @@
+import asyncio
+import contextvars
+import dataclasses
+import threading
+import weakref
+from queue import Queue
+from typing import Callable, Dict, Optional, Set, Tuple, TypeVar, Union
+
+import anyio.abc
+from typing_extensions import ParamSpec
+
+from prefect._internal.concurrency.primitives import Future
+
+T = TypeVar("T")
+P = ParamSpec("P")
+
+
+@dataclasses.dataclass
+class _WorkItem:
+    """
+    A representation of work sent to a worker thread.
+    """
+
+    future: Future
+    fn: Callable
+    args: Tuple
+    kwargs: Dict
+    context: contextvars.Context
+
+    def run(self):
+        if not self.future.set_running_or_notify_cancel():
+            return
+        try:
+            result = self.context.run(self.fn, *self.args, **self.kwargs)
+        except BaseException as exc:
+            self.future.set_exception(exc)
+            # Prevent reference cycle in `exc`
+            self = None
+        else:
+            self.future.set_result(result)
+
+
+class _WorkerThread(threading.Thread):
+    def __init__(
+        self,
+        queue: "Queue[Union[_WorkItem, None]]",  # Typing only supported in Python 3.9+
+        idle: threading.Semaphore,
+        name: str = None,
+    ):
+        super().__init__(name=name)
+        self._queue = queue
+        self._idle = idle
+
+    def run(self) -> None:
+        while True:
+            work_item = self._queue.get()
+            if work_item is None:
+                # Shutdown command received; forward to other workers and exit
+                self._queue.put_nowait(None)
```

Comment:
I'm wondering about a slightly different approach here, though I haven't thought it through. But what if `WorkerThreadPool` handed a `shutdown_event` (`Event`) to `WorkerThread` on init. The worker checks every iteration of its run loop to see if `shutdown_event` is set, and if it is, `run()` returns. I'm slightly biased toward this approach -- assuming it actually makes sense -- to avoid attaching a semantic value to `None`. Up to you though!

Reply:
Hm, the issue is that `self._queue.get()` is blocking so the worker will not do anything until it receives something in the queue. We _could_ check an event at the end of each work item but we still need to push something in the queue to wake up all the workers. Both AnyIO's worker threads and the CPython thread pool executor use this model — I trust what they're up to for now :)

Reply:
Ah, you're right -- I went looking to see how ThreadPoolExecutor handled this and missed its use of None to signal. If it's good enough for them I suppose it's good enough for us! 😂

Reply:
A bit late but I finally feel up to speed with all of the is. Double checking my understanding from https://github.com/python/cpython/blob/1332fdabbab75bc9e4bced064dc4daab2d7acb47/Lib/asyncio/queues.py#L149

`queue.get()` will only ever return `None` after an exception, so we can use it as a signal here?

Reply:
Also double checking - by blocking here, this keeps us burning through CPU on this `while True` loop, yeah?

Reply:
Yes it stops us from burning CPU. 

It returns `None` when we put `None` in the queue :D  — exceptions are raised. We place `None` in the queue to signal shutdown _and_ to wake up the worker since it is otherwise blocked.